[gazebo_ros] Changed the spawn model methods to spawn also lights.

aambrosano opened the pull request:
Created services for deleting lights, and getting and settings lights' properties.
**The code proposed in this commit is being ported from the codebase developed within the Human Brain Project, Subproject 10 (Neurorobotics), as part of the Neurorobotics Platform, by the Neurorobotics Platform team.**

aambrosano committed:
[gazebo_ros] Changed the spawn model methods to spawn also lights (and renamed accordingly).
Created services for deleting lights, and getting and settings lights' properties.

j-rivero commented:
Thanks for the pull request Alessandro. I see two different things in it:

1. Support for spawning lights
1. Rename model* fields and functions to entity*

I see no problem for merging the changes on 1 but merging the changes in 2 will break code wrote for Kinetic, which I think it is not a good idea. We were discussion about changes [in this thread]() recently. If you don't mind, I would recommend: create one pull request to implement the light feature. Another pull request to implement the change from model to entity but do not remove the \*model\* functions and messages just add the \*entity\* ones, move the functionality to there and make the old \*model\* functions to call the new \*entity\* functions together with a deprecation warning, this way we can consider them to go into Kinetic.

aambrosano committed:
[gazebo_ros] Changed the spawn model methods to spawn also lights.
Created services for deleting lights, and getting and settings lights' properties.

aambrosano committed:
[gazebo_ros] Changed the spawn model methods to spawn also lights.
Created services for deleting lights, and getting and settings lights' properties.

aambrosano committed:
[gazebo_ros] Changed the spawn model methods to spawn also lights.
Created services for deleting lights, and getting and settings lights' properties.

aambrosano committed:
[gazebo_ros] Changed the spawn model methods to spawn also lights.
Created services for deleting lights, and getting and settings lights' properties.

aambrosano commented:
Hi Jose Luis, thanks for your comments. This commit should do for the first part.

j-rivero commented on gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h:
```diff
@@ -240,21 +255,21 @@ class GazeboRosApiPlugin : public SystemPlugin
 
   /// \brief
-  void walkChildAddRobotNamespace(TiXmlNode* robot_xml);
+  void walkChildAddRobotNamespace(TiXmlNode* model_xml, std::string &model_namespace);
```
can we respect here the signature of the function that was previously in place?

j-rivero commented on gazebo_msgs/srv/SpawnModel.srv:
```diff
@@ -1,6 +1,6 @@
 string model_name                 # name of the model to be spawn
 string model_xml                  # this should be an urdf or gazebo xml
-string robot_namespace            # spawn robot and all ROS interfaces under this namespace
+string model_namespace            # spawn model and all ROS interfaces under this namespace
```
message structure and filed names are important not to break the backwards compatibility, can we keep the same name in the message than before?

aambrosano commented:
Hi Jose Luis, is there any problem with this pull request? Do you want me to remove this pull request and create a new one from scratch, or is just removing the extra code renaming functions (like I did) enough for the first step you suggested? Thank you.

j-rivero commented:
> Hi Jose Luis, is there any problem with this pull request? Do you want me to remove this pull request and create a new one from scratch, or is just removing the extra code renaming functions (like I did) enough for the first step you suggested? Thank you.

Oh, I wrote the comments about the code but forgot to press the proper "submit review" button, sorry. They should be now in place.

aambrosano committed:
Merge branch 'kinetic-devel' into kinetic-devel

davetcoleman commented on gazebo_msgs/srv/SpawnModel.srv:
```diff
@@ -1,6 +1,6 @@
 string model_name                 # name of the model to be spawn
 string model_xml                  # this should be an urdf or gazebo xml
-string model_namespace            # spawn model and all ROS interfaces under this namespace
+string robot_namespace            # spawn robot and all ROS interfaces under this namespace
```
Seems unnecessary to break a ROS message type mid-release - if I'm not mistaken this would change the md5 sum and is not API compatible. Can you just change the comment?

aambrosano replied:
The message has been put back to its original status, please check the overall changes of this pull request

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -36,6 +36,7 @@ GazeboRosApiPlugin::GazeboRosApiPlugin() :
   pub_model_states_connection_count_(0),
   pub_clock_frequency_(0)
 {
+  robot_namespace_.clear();
```
Unnecessary: default constructor "Constructs an empty string, with a length of zero characters."

aambrosano replied:
I agree, but I just put back the code that was already there, if you check the overall diff you shouldn't see any reference to robot_namespace_ at all.

davetcoleman replied:
Ah yes, I did a code review on a single commit after clicking an email link, my bad.

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -665,7 +666,7 @@ bool GazeboRosApiPlugin::spawnSDFModel(gazebo_msgs::SpawnModel::Request &req,
     // Walk recursively through the entire SDF, locate plugin tags and
     // add robotNamespace as a child with the correct namespace
-    if (!model_namespace.empty())
+    if (!this->robot_namespace_.empty())
```
no need to start using ``this->`` here

aambrosano replied:
Same here

davetcoleman commented on gazebo_ros/include/gazebo_ros/gazebo_ros_api_plugin.h:
```diff
@@ -139,14 +143,19 @@ class GazeboRosApiPlugin : public SystemPlugin
   bool deleteModel(gazebo_msgs::DeleteModel::Request &req,gazebo_msgs::DeleteModel::Response &res);
 
+  /// \brief delete a given light by name
+  bool deleteLight(gazebo_msgs::DeleteLight::Request &req,gazebo_msgs::DeleteLight::Response &res);
```
do you mind adding spaces between parameters?

aambrosano replied:
I don't, actually I would rather put them, but the line just above mine doesn't have any space between parameters.
I was wondering what the standard is when writing new functions, and it seems like throughout the code this happens many times...

davetcoleman replied:
the [roscpp style guidelines](http://wiki.ros.org/CppStyleGuide#Function_.2BAC8_Methods) don't specify this, so I guess my comment is ungrounded. but it does bother me that we are inconsistent throughout this repo on the style. as a moveit maintainer, i have setup clangformat to ensure style is automated and consistent, perhaps we should add that here (its not that hard) @j-rivero ?

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -183,6 +183,8 @@ void GazeboRosApiPlugin::loadGazeboRosApiPlugin(std::string world_name)
   gazebonode_->Init(world_name);
   //stat_sub_ = gazebonode_->Subscribe("~/world_stats", &GazeboRosApiPlugin::publishSimTime, this); // TODO: does not work in server plugin?
   factory_pub_ = gazebonode_->Advertise<gazebo::msgs::Factory>("~/factory");
+  factory_light_pub_ = gazebonode_->Advertise<gazebo::msgs::Light>("~/factory/light");
```
how about ``light_create_pub_`` so that it matches better ``light_modify_pub_``?

davetcoleman replied:
and put light at the start of the topic name too, so it matches

aambrosano replied:
This is actually made for matching the Gazebo communication channel's name, as it's happening for all the other publishers. But if you think light_create_pub_ is a better idea I will change it.

davetcoleman replied:
Ok, your call

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -219,7 +221,7 @@ void GazeboRosApiPlugin::advertiseServices()
 
   // Advertise spawn services on the custom queue
-  std::string spawn_sdf_model_service_name("spawn_sdf_model");
+  std::string spawn_sdf_model_service_name("spawn_sdf_entity");
```
this would break API right??

aambrosano replied:
Oops, this has been left back, my bad

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -517,17 +546,17 @@ void GazeboRosApiPlugin::onModelStatesDisconnect()
     ROS_ERROR_NAMED("api_plugin", "SpawnModel: Failure - model format is not URDF.");
     res.success = false;
-    res.status_message = "SpawnModel: Failure - model format is not URDF.";
+    res.status_message = "SpawnModel: Failure - entity format is not URDF.";
```
model still seems a better name for a URDF to me - it is not referring to lights here

aambrosano replied:
In both cases (spawnURDFModel and spawnSDFModel), you can pass a XML definition of a light instead of a model. That's why the comments were changed from "model" to a more generic "entity", and I initially changed the functions names as well initially (to spawnURDFEntity and spawnSDFEntity).

davetcoleman replied:
then can you reword the whole status message so that it does not just mention "URDF"?

aambrosano replied:
How? If I understand correctly the old way to define lights and the SDF one are quite different. I mean, old Gazebo versions could still have a light in a world definition, are you saying that that <light> tag is not URDF?

davetcoleman replied:
A URDF cannot specify lights, so there is not need to call the URDF an "entity" in this message unless I'm missing something.

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -584,10 +613,10 @@ bool GazeboRosApiPlugin::spawnURDFModel(gazebo_msgs::SpawnModel::Request &req,
                                        gazebo_msgs::SpawnModel::Response &res)
 {
-  // incoming robot name
+  // incoming entity name
```
same comment

aambrosano replied:
Same here

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -747,6 +777,40 @@ bool GazeboRosApiPlugin::deleteModel(gazebo_msgs::DeleteModel::Request &req,
+{
+  gazebo::physics::LightPtr phy_light = world_->Light(req.light_name);
+
+  if (phy_light == NULL) {
```
ROS style guidelines have line break for brackets

aambrosano replied:
I'm going to fix this in the next patchset

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -747,6 +777,40 @@ bool GazeboRosApiPlugin::deleteModel(gazebo_msgs::DeleteModel::Request &req,
+
+  if (phy_light == NULL) {
+    res.success = false;
+    res.status_message = std::string("DeleteLight: Requested light ") + req.light_name + std::string(" not found!");
```
is ``std::string`` necessary for all the concatenations?

aambrosano replied:
I thought it was, but I was wrong.

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -2014,10 +2139,10 @@ void GazeboRosApiPlugin::updateSDFAttributes(TiXmlDocument &gazebo_model_xml, st
   if (model_tixml)
   {
-    // Update model name
+    // Update entity name
```
model still seems like better comment here

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -2266,16 +2397,35 @@ bool GazeboRosApiPlugin::spawnAndConform(TiXmlDocument &gazebo_model_xml, std::s
   gazebo::physics::ModelPtr model = world_->GetModel(model_name);
-  if (model)
+  gazebo::physics::LightPtr light = world_->Light(model_name);
+  if ((isLight && light != NULL) || (model != NULL))
```
is it possible there is both a model and a light with the same name? this conditional would not catch that error

aambrosano replied:
I don't need to catch that particular case, what is being done here is:
- if the entity to spawn is a light, make sure there's no light with the same name
- if it's a model, then check there's no model with the same name
All the other cases should be fine. Lights and models with the same name are allowed AFAIK, as they're saved in different data structures.

davetcoleman commented on gazebo_ros/src/gazebo_ros_api_plugin.cpp:
```diff
@@ -2266,16 +2397,35 @@ bool GazeboRosApiPlugin::spawnAndConform(TiXmlDocument &gazebo_model_xml, std::s
 
-  // Publish the factory message
-  factory_pub_->Publish(msg);
+  #if GAZEBO_MAJOR_VERSION > 6
```
i think we shouldn't support versions older than 6 for the kinetic branch, see https://github.com/ros-simulation/gazebo_ros_pkgs/pull/542

aambrosano replied:
Ok, though there's some "#if GAZEBO_MAJOR_VERSION >= 3" around, I think they should be removed as well. Should I do this in this same pull request?

davetcoleman replied:
Different pull request, thanks for helping with code maintenance!

davetcoleman reviewed (changes requested):
Great contribution, I almost always have to add a light manually when I launch Gazebo so I can see whats going on in the dark environment!

davetcoleman commented:
i just noticed @j-rivero comments, and i agree with them

davetcoleman commented:
@aambrosano I'm excited to use this feature, do you have example roslaunch code for adding a light to the scene?

aambrosano commented:
@davetcoleman I didn't prepare any roslaunch code but you can try like this:
- Run `rosrun gazebo_ros gazebo`
- Copy this code on your shell
<pre>
rosservice call /gazebo/spawn_sdf_model "model_name: 'dirlight1'
model_xml: '&lt;sdf version="1.5"&gt;
    &lt;light type="directional"&gt;
      &lt;pose frame=""&gt;2.32048 -2.18175 1 0 -0 0&lt;/pose&gt;
      &lt;diffuse&gt;0.5 0.5 0.5 1&lt;/diffuse&gt;
      &lt;specular&gt;0.1 0.1 0.1 1&lt;/specular&gt;
      &lt;direction&gt;0.1 0.1 -0.9&lt;/direction&gt;
      &lt;attenuation&gt;
        &lt;range&gt;20&lt;/range&gt;
        &lt;constant&gt;0.5&lt;/constant&gt;
        &lt;linear&gt;0.01&lt;/linear&gt;
        &lt;quadratic&gt;0.001&lt;/quadratic&gt;
      &lt;/attenuation&gt;
      &lt;cast_shadows&gt;1&lt;/cast_shadows&gt;
    &lt;/light&gt;
&lt;/sdf&gt;'
robot_namespace: ''
initial_pose:
  position: {x: 0.0, y: 0.0, z: 0.0}
  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
</pre>

Sorry for the bad readability of this example, let me know if it works!

davetcoleman commented:
The rosservice call is not working for me... first I had to change ``entity`` back to ``model`` but still I get errors using the updated command:

```
rosservice call /gazebo/spawn_sdf_model "entity_name: 'dirlight1'
model_xml: '<sdf version="1.5">
<light type="directional">
<pose frame="">2.32048 -2.18175 1 0 -0 0</pose>
<diffuse>0.5 0.5 0.5 1</diffuse>
<specular>0.1 0.1 0.1 1</specular>
<direction>0.1 0.1 -0.9</direction>
<attenuation>
<range>20</range>
<constant>0.5</constant>
<linear>0.01</linear>
<quadratic>0.001</quadratic>
</attenuation>
<cast_shadows>1</cast_shadows>
</light>
</sdf>'
robot_namespace: ''
initial_pose: 
position: {x: 0.0, y: 0.0, z: 0.0}
orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
```

Error:
```
ERROR: Incompatible arguments to call service:
No field name [orientation]

Service arguments are: [model_name model_xml robot_namespace initial_pose.position.x initial_pose.position.y initial_pose.position.z initial_pose.orientation.x initial_pose.orientation.y initial_pose.orientation.z initial_pose.orientation.w reference_frame]
```

I'm not sure how to nestle the ``initial_pose`` group.

aambrosano commented:
That's right, I forgot to change the field name (I was trying with my local version yesterday when I wrote you, my bad). I think adding some spaces before pose and orientation should be fine. I'll try to update the comment accordingly so you can copy directly from there.

aambrosano commented:
@davetcoleman did you manage to test the changes?

davetcoleman commented:
@aambrosano which changes? did you update a comment? could you just repost the correct version - comment updates don't trigger emails fyi

aambrosano commented:
@davetcoleman sure, sorry for the inconvenient.
The code to test the service is this:
<pre>
rosservice call /gazebo/spawn_sdf_model "model_name: 'dirlight1'
model_xml: '&lt;sdf version="1.5"&gt;
    &lt;light type="directional"&gt;
      &lt;pose frame=""&gt;2.32048 -2.18175 1 0 -0 0&lt;/pose&gt;
      &lt;diffuse&gt;0.5 0.5 0.5 1&lt;/diffuse&gt;
      &lt;specular&gt;0.1 0.1 0.1 1&lt;/specular&gt;
      &lt;direction&gt;0.1 0.1 -0.9&lt;/direction&gt;
      &lt;attenuation&gt;
        &lt;range&gt;20&lt;/range&gt;
        &lt;constant&gt;0.5&lt;/constant&gt;
        &lt;linear&gt;0.01&lt;/linear&gt;
        &lt;quadratic&gt;0.001&lt;/quadratic&gt;
      &lt;/attenuation&gt;
      &lt;cast_shadows&gt;1&lt;/cast_shadows&gt;
    &lt;/light&gt;
&lt;/sdf&gt;'
robot_namespace: ''
initial_pose:
  position: {x: 0.0, y: 0.0, z: 0.0}
  orientation: {x: 0.0, y: 0.0, z: 0.0, w: 0.0}
reference_frame: ''"
</pre>

davetcoleman commented:
Just tested and it works for me, thanks!

Please add a [documentation section here](http://gazebosim.org/tutorials?tut=ros_roslaunch) under spawn model that has your example for spawning lights
